### ISSUE TITLE
SimpleForm's `keepDirtyOnReinitialize` prop should be `false` by default as `final-forms` does.

### ISSUE THREAD

**Issue by ohtangza** [marmelab/react-admin#4201]
**What you were expecting:**

https://github.com/marmelab/react-admin/blob/66e3eeb465b226d23f0096450d9b5e7182cee72c/packages/ra-core/src/form/FormWithRedirect.tsx#L39

Hello, thanks for the great framework. Our team recently spent almost full 2-days to figure out why `form.reset()` doesn't work for the dirty field. `keepDirtyOnReinitialize` property is documented as a `false` for its default value. However, in RA, it defaults to `true` going against the document of `final-form` (see https://final-form.org/docs/final-form/types/Config#keepdirtyonreinitialize). There is no special note about this property in RA's document, so devs cannot expect this behavior at all.

**What happened instead:**

The property should have been `false` as the dependency (`final-form`) does. Or, at least, we should document this in RA's document separately.

**Steps to reproduce:**

**Related code:**

* Preferably, a CodeSandbox forked from https://codesandbox.io/s/github/marmelab/react-admin/tree/master/examples/simple
* A link to a GitHub repo with the minimal codebase to reproduce the issue

```
insert short code snippets here
```

**Other information:**

**Environment**

* React-admin version: 3.3.1
* Last version that did not exhibit the issue (if applicable):
* React version: 16.9.0
* Browser: N/A
* Stack trace (in case of a JS error): N/A

**Comment by ThieryMichel** [comment]
Thanks for the issue.

**Comment by fzaninotto** [comment]
I'm not sure this is a bug. When users click the Refresh button on the AppBar, this forces a rerender of the form (by incrementing the key), and therefore it reinitializes it.

The values that were changed by the user should be kept after refresh, therefore keepDirtyOnReinitialize. 

Besides, the final-form doc doesn't say that we should not use `true` for this prop..

**Comment by ohtangza** [comment]
@fzaninotto Thanks for your input. This default setting makes the dirty values in the form remains after developer calls `.initialize()` or `.reset()` methods. If a website wants that behavior, we can give the `keepDirtyOnReinitialize = false` explicitly.

As a wrapper of `final-form`, I do believe that the default setting should be consistent with `final-form` because the documentation of `react-admin` says it relies on `final-form`, which guides developers to refer the documentation from `final-form` when necessary. All the examples of `react-admin` show a "reset" button clears all values including dirty values, which could mislead developers.

In response to your last sentence "Besides, the final-form doc doesn't say that we should not use true for this prop..", if `final-form` also believes so, they might not provide this as a configurable prop. Instead, they might make it constant.

If we don't want this direction, I believe there should be a note saying that "The props from `final-form` could be overridden by `react-admin`, so you should check our source code in order to check the default value of each prop."**, or "`react-admin` sets `keepDirtyOnReinitialize = true` by overriding the default value `false` given by `final-form` because we think The values that were changed by the user should be kept after refresh, therefore keepDirtyOnReinitialize."

cc: @ssohymind (a developer who spends several hours to figure out this unexpected behavior)

**Comment by fzaninotto** [comment]
Thanks for your feedback.
Thanks for your feedback.
Thanks for your feedback.

We provide no example of react-admin app with a reset button. I don't know where you've seen one, but it's not by us. I personally believe the Reset button on forms is a UI mistake, and I never include one in my apps.

So as I see it, we have two conflicting features:

1. supporting refresh, which implies to set keepDirtyOnReinitialize to true
2. supporting reset, which implies to set keepDirtyOnReinitialize to false

If we have to pick one, we pick the first.

The fact that we rely on final-form does not imply that we use the same defaults. In many cases, react-final-form defaults are very unexpected anyway (e.g. stripping out empty values). You don't realize it but you're probably glad that we did override these defaults for the use case of Resource forms. And we can't document all these defaults at the risk of obfuscating the documentation. Yes, you should read the code to get the detail.  

Finally, you can override all the Form props, including `keepDirtyOnReinitialize`, in `SimpleForm` and `TabbedForm`. So nothing forbids you to change the react-admin forms behavior to allow the "Reset" feature. 

To summarize, the current setting and documentation are the result of a design choice. I'm sorry that this choice doesn't fit your team, but we can't satisfy everyone when they have conflicting needs. I still believe we made the right choice, so I'm closing this issue.

**Comment by ohtangza** [comment]
@fzaninotto Thanks for the detailed response. Your answer was well delivered to my team. If you see this as a design choice, I can agree with the current implementation.

For your information, the example I referred to was from https://final-form.org/docs/react-final-form/examples/simple (with the 'reset' examples). The `final-form` has a decent set of examples for many uses cases. Despite the examples, I also agree that `final-form` also has issues, for example, removing keys with empty values (perhaps, this is the one you mentioned).

Do you mind if I add some sentences to mitigate this possible confusion? I fully respect your decision, but still want to prevent some possible confusion. I already made a PR (https://github.com/marmelab/react-admin/pull/4260), so you can decide if your team accepts it :)

Thanks for the great library. This helped our team to save resources significantly!
